### ISSUE TITLE
Disable unnecessary-dunder-call for __init__

### ISSUE THREAD

**Issue by felipe-olmos-orange** [PyCQA/pylint#7830]
### Current problem

Most of the time `unnecessary-dunder-call` is desirable but not with the `__init__` method.

A legitimate use of calling `__init__` is when you have no-parameter (or only keyword params) constructor. If somewhere in the code you want to reinitialize the object to its default state one just need to call `__init__`. The enforcing of this rule either forces you to create custom `_my_init` methods to be called by `__init__` or use new objects which are a waste of memory.

### Desired solution

`unnecessary-dunder-call` should be disabled for `__init__`

**Comment by clavedeluna** [comment]
@felipe-olmos-orange can you add a code example for what you're talking about?

**Comment by felipe-olmos-orange** [comment]
@clavedeluna Here's a quick one

```python

class Runner:
    def __init__(output_dir="/tmp"):
        self.output_dir = output_dir
    
    def run():
        # Run some process using output_dir for its output
        pass

# Somewhere in the code        
my_runner = Runner()
my_runner.output_dir = "/my/output/dir"
my_runner.run()
# Reinitialize to the default state, whatever it is, this could be done even later
my_runner.__init__()
```

Note that the state in this example is trivial, but it could be much more complex.  A `reinitialize` method would duplicate `__init__`.

**Comment by clavedeluna** [comment]
I think that code could genuinely be refactored to use a pattern that doesn't require calling init again. It feels off to me but I can't quite scramble to find the right pattern to make this code pattern "better".  If anything, it seems like everything under init could be called with `self.reset()` or `self.initialize() or even `self.update()`

**Comment by felipe-olmos-orange** [comment]
For me that `__init__` has the semantic meaning of "(re)initialization", and I don't think unreasonable to avoid maintaining a `reset` method every time this pattern is needed.

Apart that stack overflow discussion and that of [this feature](https://github.com/PyCQA/pylint/pull/5938), it is assumed that is a bad practice without much more discussion.

**Comment by jacobtylerwalls** [comment]
Try calling init on the class, not the instance. I thought the way the feature was written we allowed calling init on the class (or on super).

**Comment by jacobtylerwalls** [comment]
Oh, sorry. I looked at the use case more carefully. I see you're trying to avoid the overhead of creating a new instance? I'm in favor of no change and just disabling this if wanted.

**Comment by Pierre-Sassoulas** [comment]
I'm closing as not planned, as 4 maintainers and stackoverflow seems to agree on this.
